File: homework01/vigenere.py
def encrypt_vigenere(plaintext: str, keyword: str) -> str:
    """
    Encrypts plaintext using a Vigenere cipher.

    >>> encrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> encrypt_vigenere("python", "a")
    'python'
    >>> encrypt_vigenere("ATTACKATDAWN", "LEMON")
    'LXFOPVEFRNHR'
    """

    ciphertext = ""

    if len(plaintext) > len(keyword):
        key = (keyword * (1 + (len(plaintext) // len(keyword)))).lower()
    else:
        key = keyword.lower()

    shifts = []
    k = 0

    for i in key:
        k += 1
        if k <= len(plaintext):
            shifts.append(ord(i) - ord("a"))

    for i, m in zip(plaintext, shifts):
        if ord("A") <= ord(i) <= ord("Z") - m:
            ciphertext += chr(ord(i) + m)
        elif ord("a") <= ord(i) <= ord("z") - m:
            ciphertext += chr(ord(i) + m)
        elif ord("Z") - m <= ord(i) <= ord("a"):
            ciphertext += chr(ord("A") - 1 + m - ord("Z") + ord(i))
        elif ord("a") - m <= ord(i) <= ord("z"):
            ciphertext += chr(ord("a") - 1 + m - ord("z") + ord(i))
        else:
            ciphertext += i

    return ciphertext


def decrypt_vigenere(ciphertext: str, keyword: str) -> str:
    """
    Decrypts a ciphertext using a Vigenere cipher.

    >>> decrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> decrypt_vigenere("python", "a")
    'python'
    >>> decrypt_vigenere("LXFOPVEFRNHR", "LEMON")
    'ATTACKATDAWN'
    """
    plaintext = ""

    if len(ciphertext) > len(keyword):
        key = (keyword * (1 + (len(ciphertext) // len(keyword)))).lower()
    else:
        key = keyword.lower()

    shifts = []
    k = 0

    for i in key:
        k += 1
        if k <= len(ciphertext):
            shifts.append(ord(i) - ord("a"))

    for i, m in zip(ciphertext, shifts):
        if ord("A") + m <= ord(i) <= ord("Z"):
            plaintext += chr(ord(i) - m)
        elif ord("a") + m <= ord(i) <= ord("z"):
            plaintext += chr(ord(i) - m)
        elif ord("A") <= ord(i) <= ord("A") + m:
            plaintext += chr(ord("Z") + 1 - abs(ord(i) - m - ord("A")))
        elif ord("a") <= ord(i) <= ord("a") + m:
            plaintext += chr(ord("z") + 1 - abs(ord(i) - m - ord("a")))
        else:
            plaintext += i

    return plaintext

File: homework01-go/vigenere/vigenere.go
package vigenere

import (
	"math"
	"strings"
)

func EncryptVigenere(plaintext string, keyword string) string {

	const Max_ABC = int('Z')
	const Min_ABC = int('A')
	const Max_abc = int('z')
	const Min_abc = int('a')

	var key string

	if len(plaintext) > len(keyword) {
		key = strings.ToLower(strings.Repeat(keyword, (1 + len(plaintext)/len(keyword))))
	} else {
		key = strings.ToLower(keyword)
	}

	var shifts []int

	for i := 0; i < len(key); i++ {
		shifts = append(shifts, (int(key[i]) - int('a')))
	}

	var ciphertext string

	for i := 0; i < len(plaintext); i++ {
		var plain_letter int = int(plaintext[i])
		var shift int = int(shifts[i])

		if plain_letter >= Min_ABC && plain_letter <= Max_ABC-shifts[i] {
			ciphertext += string(rune(plain_letter + shift))
		} else if plain_letter >= Min_abc && plain_letter <= Max_abc-shift {
			ciphertext += string(rune(plain_letter + shift))
		} else if plain_letter >= Max_ABC-shift && plain_letter <= Min_abc {
			ciphertext += string(rune(Min_ABC - 1 + shift - Max_ABC + plain_letter))
		} else if plain_letter >= Min_abc-shift && plain_letter <= Max_abc {
			ciphertext += string(rune(Min_abc - 1 + shift - Max_abc + plain_letter))
		} else {
			ciphertext += string(rune(plain_letter))
		}
	}
	return ciphertext
}

func DecryptVigenere(ciphertext string, keyword string) string {

	const Max_ABC = int('Z')
	const Min_ABC = int('A')
	const Max_abc = int('z')
	const Min_abc = int('a')

	var key string

	if len(ciphertext) > len(keyword) {
		key = strings.ToLower(strings.Repeat(keyword, (1 + len(ciphertext)/len(keyword))))
	} else {
		key = strings.ToLower(keyword)
	}

	var shifts []int

	for i := 0; i < len(key); i++ {
		shifts = append(shifts, (int(key[i]) - int('a')))
	}

	var plaintext string

	for i := 0; i < len(ciphertext); i++ {
		var ciph_letter int = int(ciphertext[i])
		var shift int = int(shifts[i])

		if ciph_letter >= Min_ABC+shift && ciph_letter <= Max_ABC {
			plaintext += string(rune(ciph_letter - shift))
		} else if ciph_letter >= Min_abc+shift && ciph_letter <= Max_abc {
			plaintext += string(rune(ciph_letter - shift))
		} else if ciph_letter >= Min_ABC && ciph_letter <= Min_ABC+shift {
			plaintext += string(rune(Max_ABC - int(math.Abs(float64(ciph_letter-shift-Min_ABC))) + 1))
		} else if ciph_letter >= Min_abc && ciph_letter <= Min_abc+shift {
			plaintext += string(rune(Max_abc - int(math.Abs(float64(ciph_letter-shift-Min_abc))) + 1))
		} else {
			plaintext += string(rune(ciph_letter))
		}
	}
	return plaintext
}
